python_modules/dagster/dagster/_core/execution/context/compute.py

```diff
@@ -1263,3 +1267,59 @@ def set_requires_typed_event_stream(self, *, error_message: Optional[str] = None
 class AssetExecutionContext(OpExecutionContext):
     def __init__(self, step_execution_context: StepExecutionContext):
         super().__init__(step_execution_context=step_execution_context)
+
+
+def build_execution_context(
+    step_context: StepExecutionContext,
+) -> Union[OpExecutionContext, AssetExecutionContext]:
+    """Get the correct context based on the type of step (op or asset) and the user provided context
+    type annotation. Follows these rules.
+
+    step type     annotation                result
+    asset         AssetExecutionContext     AssetExecutionContext
+    asset         OpExecutionContext        OpExecutionContext
+    asset         None                      AssetExecutionContext
+    op            AssetExecutionContext     Error - we cannot init an AssetExecutionContext w/o an AssetsDefinition
+    op            OpExecutionContext        OpExecutionContext
+    op            None                      OpExecutionContext
+    For ops in graph-backed assets
+    step type     annotation                result
+    op            AssetExecutionContext     AssetExecutionContext
+    op            OpExecutionContext        OpExecutionContext
+    op            None                      OpExecutionContext
+    """
+    is_sda_step = step_context.is_sda_step
+    is_op_in_graph_asset = is_sda_step and step_context.is_op_in_graph
+    context_annotation = EmptyAnnotation
+    compute_fn = step_context.op_def._compute_fn  # noqa: SLF001
+    compute_fn = (
+        compute_fn
+        if isinstance(compute_fn, DecoratedOpFunction)
+        else DecoratedOpFunction(compute_fn)
+    )
+    if compute_fn.has_context_arg():
+        context_param = compute_fn.get_context_arg()
+        context_annotation = context_param.annotation
+
+    # TODO - i dont know how to move this check to Definition time since we don't know if the op is
+    # part of a graph-backed asset until we have the step execution context, i think
+    if context_annotation is AssetExecutionContext and not is_sda_step:
```

Comment:
It'd be nice to move this check to definition time, but i dont think we can. At definition time, I dont know of a way to know if 
```
@op 
def my_op(context: AssetExecutionContext)
```
is being used in a job or in a graph-backed asset

Reply:
i think this makes sense

nit: would reframe the comment before land to something like "would be nice to ... currently doesnt seem possible"

Reply:
ok cool, will update